### ISSUE TITLE
THCTensorCopy.c error (77)

### ISSUE THREAD

**Issue by hworang77** [torch/cutorch#335]
Hi, I got the following error 

`/home/paul/torch/install/bin/luajit: cuda runtime error (77) : an illegal memory access was encountered at /home/paul/torch/extra/cutorch/lib/THC/generic/THCStorage.c:147`

After adding the code at #330 the error message says it occurred while copying which actually matches the code I found by printing intermediate results in my code.

`THCudaCheck FAIL file=/home/paul/torch/extra/cutorch/lib/THC/generic/THCTensorCopy.cu line=74 error=77 : an illegal memory access was encountered
/home/paul/torch/install/bin/luajit: cuda runtime error (77) : an illegal memory access was encountered at /home/paul/torch/extra/cutorch/lib/THC/generic/THCStorage.c:147
`

But interestingly, the exact same code works with the same data without any error when I change the size of intermediate layers of my network (it works even when the network size is bigger than the current setting). Because of this I thought that the error doesn't come from passing out-of-bound indices (even to the network with different size, the same indices were passed).

The code causing the error is as follows:
`drnn_out:copy(drnn_state[t+1])`
here `drnn_out` is a cudatensor of size 100 * 2048 and `drnn_state[t+1]` is a table of cudatensors of size 100 * 2048.
`t` changes from the `#drnn_state-1` to 1 the error occurs when `t` becomes `#drnn_state-2` (which means when it was `#drnn_state-1`, this code worked). when I tried to remove this code then the error occurs at the code right after that which is
`drnn_out[k] = dmult[1][k]`

I have been struggling with this for a whole day, any help please?

**Comment by soumith** [comment]
run your program with:

CUDA_LAUNCH_BLOCKING=1 th [yourprogram.lua]

After that, you'll find the exact place where the issue is occuring.

**Comment by hworang77** [comment]
sadly, I've already been doing that...

I got those messages using that flag

**Comment by soumith** [comment]
if you have a small ~100 line repro, i'm happy to have a look. otherwise, you might have to give cuda-gdb a try...

**Comment by hworang77** [comment]
`function feval()
   embed_grad:zero()
   rnn_grad:zero()
   mult_grad:zero()

   ------------------- get minibatch -------------------
   local feats, x, x_len, y, qids = vqa_data.train_data:next_batch_feat(opt.feat_dir, cnnout_dim)
   if opt.gpuid >= 0 then
      feats = feats:float():cuda()
      x = x:float():cuda()
      y = y:float():cuda()
   end

   -- make sure we are in correct omde (this is cheap, sets flag)
   protos.multimodal:training()
   -------------------- forward pass -------------------
   local loss = 0
   -- RNN FORWARD
   local max_len = x_len:max()
   local min_len = x_len:min()
   local we_vecs = {}
   local rnn_state = {[0] = init_state}
   rnn_out:zero()
   for t = 1, max_len do
      mult_protos.rnns[t]:training()
      mult_protos.word_embeds[t]:training()
      local we = mult_protos.word_embeds[t]:forward(x[{t,{}}])
      local lst = mult_protos.rnns[t]:forward({we, rnn_state[t-1]})
      we_vecs[t] = we
      rnn_state[t] = lst
      if t >= min_len then
         for k = 1, opt.batch_size do
            if x_len[k] == t then
               rnn_out[k] = lst[k]
            end
         end
      end
   end
   -- MULTIMODAL FORWARD
   local pred = protos.multimodal:forward({rnn_out, feats})

   -- compute accuracy
   local max_score, ans = torch.max(pred, 2)
   ans = torch.squeeze(ans)
   train_acc = train_acc + torch.eq(ans, y):sum()
   train_num_data = train_num_data + y:nElement()
   -- loss forward
   loss = loss + protos.criterion:forward(pred, y)

   -------------------- backward pass -------------------
   -- loss backward
   local dpred = protos.criterion:backward(pred, y)
   local dmult = protos.multimodal:backward({rnn_out, feats}, dpred)

   -- rnn backward
   local drnn_state = {[max_len+1] = init_state:clone()}
                    -- true also zeros the clones
   for t = max_len, 1, -1 do
      drnn_out:copy(drnn_state[t+1])  -- this line causes the error
      if t >= min_len then
         for k = 1, opt.batch_size do
            if x_len[k] == t then
               drnn_out[k] = dmult[1][k]
            end
         end
      end

      local dlst = mult_protos.rnns[t]:backward({we_vecs[t], rnn_state[t-1]}, drnn_out)
      local dwe = mult_protos.word_embeds[t]:backward(x[{t, {}}], dlst[1])

      drnn_state[t] = dlst[2] -- dlst[1] is gradient on x, which we don't need
   end`

**Comment by soumith** [comment]
i cant actually run this

**Comment by hworang77** [comment]
ah, okay i didn't actually understand your comment.

hmm making a small reproduction would be quiet hard since I have been using torch quiet long and this is the first time to face this error. Would give me a hint to run cuda-gdb and what to see?

**Comment by soumith** [comment]
this error only usually occurs for me when I have a ClassNLLCriterion, and I give a target class that is bigger than nClasses. Check for that case. if that's not your issue, cuda-gdb pointer is: http://docs.nvidia.com/cuda/cuda-gdb/ use it like regular gdb and debug.

**Comment by hworang77** [comment]
While I was testing by removing lines one by one to reduce the size of the code, i found that the code caused the error after reading hdf5.
